app/test/shared/test_services.dart

```diff
@@ -95,13 +99,42 @@ void testWithServices(String name, Future fn()) {
   });
 }
 
-http_testing.MockClientHandler _wrapShelfHandler(shelf.Handler handler) {
+/// Creates local, non-HTTP-based API client with [authToken].
+PubApiClient createPubApiClient({String authToken}) =>
+    PubApiClient('http://localhost:0/',
+        client: httpClient(authToken: authToken));
+
+/// Returns a HTTP client that bridges HTTP requests and shelf handlers without
+/// the actual HTTP transport layer.
+///
+/// If [handler] is not specified, it will use the default frontend handler.
+http_testing.MockClient httpClient({
+  shelf.Handler handler,
+  String authToken,
+}) {
+  handler ??= createAppHandler(null);
+  handler = wrapHandler(
+    Logger.detached('test'),
+    handler,
+    sanitize: true,
+  );
+  return http_testing.MockClient(
+      _wrapShelfHandler(handler, authToken: authToken));
```

Comment:
Probably unrelated to this PR, just curious why don't we listen on a port an make actual http requests, instead of using MockClient, which seems like it's a fake client?

Reply:
It is much faster to do this wrapping than waiting for the OS to open an unused port number. I think with the current wrapping we cover a lot in our code, and using the HTTP wrapping wouldn't cover more of it.